File: TOSAi.Mobile/lib/core/api/fallback_platform_api_client.dart
import '../models/platform_feature_page.dart';
import '../models/user_role.dart';
import 'platform_api_client.dart';

class FallbackPlatformApiClient implements PlatformApiClient {
  const FallbackPlatformApiClient({
    required this.primary,
    required this.fallback,
  });

  final PlatformApiClient primary;
  final PlatformApiClient fallback;

  @override
  Future<PlatformFeaturePage> getFeaturePage({
    required UserRole role,
    required String pageKey,
  }) async {
    try {
      return await primary.getFeaturePage(role: role, pageKey: pageKey);
    } catch (_) {
      return fallback.getFeaturePage(role: role, pageKey: pageKey);
    }
  }
}

File: TOSAi.TeacherApp/Services/FallbackPlatformApiClient.cs
using TOSAi.TeacherApp.Models;

namespace TOSAi.TeacherApp.Services;

public sealed class FallbackPlatformApiClient : IPlatformApiClient
{
    private readonly IPlatformApiClient _primary;
    private readonly IPlatformApiClient _fallback;

    public FallbackPlatformApiClient(IPlatformApiClient primary, IPlatformApiClient fallback)
    {
        _primary = primary;
        _fallback = fallback;
    }

    public async Task<PlatformFeaturePage> GetFeaturePageAsync(UserRole role, string pageKey, CancellationToken cancellationToken = default)
    {
        try
        {
            return await _primary.GetFeaturePageAsync(role, pageKey, cancellationToken);
        }
        catch
        {
            return await _fallback.GetFeaturePageAsync(role, pageKey, cancellationToken);
        }
    }
}
